### PR TITLE
iRobotCreate

### PR DESCRIPTION
**Pull request by chrismbirmingham**
This pull request adds the iRobotCreate to the simulator. The create is defined using the same Robot interface as the demobot, but makes use of the create IO library for motor control and sensor reading. 

Additionally this pull request has moved files for better organization, documentation, and bug fixes.

### CODE REVIEW

File: dependencies/build.py [hunk 1]
```diff
@@ -244,9 +244,9 @@ def is_tool(name):
   'libkipr_c': f'{libkipr_install_c_dir}',
   'libkipr_python': f'{libkipr_build_python_dir}',
   'cpython': f'{cpython_emscripten_build_dir}',
-  'ammo': f'{ammo_build_dir}',
+  # 'ammo': f'{ammo_build_dir}',
```
Comment (bmcdorman) [inline dependencies/build.py:247]: delete line rather than comment

File: express.js [hunk 2]
```diff
@@ -118,6 +118,7 @@ if (config.server.dependencies.libkipr_c && config.server.dependencies.emsdk_env
                   error: `Failed to delete ${path}`
                 });
               }
+              // console.log("Result successfully sent", data.toString());
```
Comment (bmcdorman) [inline express.js:121]: nit: delete

File: src/BabylonBindings/RobotObjects/CreateBindings.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+import Geometry from "../../state/State/Scene/Geometry";
+import Sensor from '../Sensors/Sensor';
+
+import { SensorPacket } from './SensorPacket';
+import { BuiltGeometry } from './RobotLink';
+import { RawVector3 } from '../../util/math/math';
+
+
+
+class CreateBinding {
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/CreateBindings.ts:26]: Why is this file not named `CreateBinding`?

File: src/BabylonBindings/RobotObjects/MotorBindings.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import { Scene as babylScene, Quaternion, Vector3, Mesh,
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/MotorBindings.ts:1]: This file should have a `camelCase` name if it is not a default class export

File: src/BabylonBindings/RobotObjects/MotorBindings.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * 
+ * @param id - The ID of the hinge joint. A string that likely uniquely identifies the hinge joint. 
+ * @param hinge - The hinge joint object containing parent and child information. An object that contains information about the hinge joint, 
+ *          including the parent's pivot point, the child's pivot point, and the axes of rotation for both the parent and child.
+ * @param bScene_ - The Babylon scene in which the parent and child objects exist.
+ * @param bParent - The parent mesh.
+ * @param bChild - The child mesh.
+ * @returns The created 6DoF constraint.
+ */
+export const createHinge_ = (id: string, hinge: Node.HingeJoint & { parentId: string }, bScene_: babylScene, bParent: Mesh, bChild: Mesh) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/MotorBindings.ts:26]: The `_` suffix is only for private class members

File: src/BabylonBindings/RobotObjects/RobotLink.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  export interface Collider {
+    name: string;
+    mesh: Mesh;
+    type: number;
+    volume: number;
+  }
+}
+
+// Loads the geometry of a robot part and divides into the collider and noncollider pieces
+export const buildGeometry_ = async (name: string, geometry: Geometry, bScene_: babylScene): Promise<BuiltGeometry> => {
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/RobotLink.ts:28]: The `_` suffix is only for private class members.

File: src/BabylonBindings/RobotObjects/RobotLink.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    default: { throw new Error(`Unsupported geometry type: ${geometry.type}`); }
+  }
+  return ret;
+};
+
+
+
+// Creates a link and returns the root mesh. Links are wheels, chasis, arms, etc.
+// Loads the geometry and adds the appropriate physics properties to the mesh
+export const createLink_ = async (id: string, link: Node.Link, bScene_: babylScene, robot_: Robot, colliders_: Set<Mesh>) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/RobotLink.ts:75]: The `_` suffix is only for private class members.

File: src/BabylonBindings/RobotObjects/WeightBinding.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/WeightBinding.ts:1]: File name should be `camelCase`

File: src/BabylonBindings/RobotObjects/WeightBinding.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import Node from '../../state/State/Robot/Node';
+import { ReferenceFramewUnits } from '../../util/math/UnitMath';
+import { RENDER_SCALE } from '../../components/Constants/renderConstants';
+import { Mass } from '../../util/math/Value';
+import Robot from '../../state/State/Robot';
+import Dict from '../../util/objectOps/Dict';
+
+// Adds an invisible weight to a parent link.
+export const createWeight_ = (id: string, weight: Node.Weight, bScene_: babylScene, robot_: Robot, links_: Dict<Mesh>) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/RobotObjects/WeightBinding.ts:14]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/CameraBinding.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import {  ArcRotateCamera, Scene as babylScene, Camera as babylCamera } from '@babylonjs/core';
+
+import Camera from "../../state/State/Scene/Camera";
+import { Vector3wUnits } from "../../util/math/UnitMath";
+
+
+
+export const createArcRotateCamera = (camera: Camera.ArcRotate, bScene_: babylScene): ArcRotateCamera => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/CameraBinding.ts:8]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/CameraBinding.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  const ret = new ArcRotateCamera('botcam', 0, 0, 0, Vector3wUnits.toBabylon(camera.target, 'centimeters'), bScene_);
+  ret.attachControl(bScene_.getEngine().getRenderingCanvas(), true);
+  ret.position = Vector3wUnits.toBabylon(camera.position, 'centimeters');
+  ret.panningSensibility = 100;
+  // ret.checkCollisions = true;
+  return ret;
+};
+
+export const createNoneCamera = (camera: Camera.None, bScene_: babylScene): ArcRotateCamera => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/CameraBinding.ts:17]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/Lightbinding.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
+import { ShadowGenerator,  IShadowLight, PointLight, SpotLight, DirectionalLight,
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/Lightbinding.ts:1]: This file name should be `camelCase`. Currently it obeys no naming scheme at all.

File: src/BabylonBindings/SceneObjects/Lightbinding.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    RawVector3.toBabylon(node.direction),
+    Angle.toRadiansValue(node.angle),
+    node.exponent,
+    bScene_
+  );
+  shadowGenerators_[id] = createShadowGenerator_(ret);
+  return ret;
+};
+
+export const createPointLight = (id: string, node: Node.PointLight, bScene_: babylScene, shadowGenerators_: Dict<ShadowGenerator>): PointLight => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/Lightbinding.ts:37]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/Lightbinding.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+import Node from "../../state/State/Scene/Node";
+import Dict from "../../util/objectOps/Dict";
+import { ReferenceFramewUnits, Vector3wUnits } from "../../util/math/UnitMath";
+import { RawVector3 } from "../../util/math/math";
+import LocalizedString from '../../util/LocalizedString';
+import { Angle } from "../../util";
+
+
+export const createDirectionalLight = (id: string, node: Node.DirectionalLight, bScene_: babylScene, shadowGenerators_: Dict<ShadowGenerator>): DirectionalLight => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/Lightbinding.ts:12]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/Lightbinding.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    bScene_
+  );
+  ret.intensity = node.intensity;
+  
+  shadowGenerators_[id] = createShadowGenerator_(ret);
+  ret.setEnabled(node.visible);
+  return ret;
+};
+
+const createShadowGenerator_ = (light: IShadowLight) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/Lightbinding.ts:52]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/MaterialBinding.ts [hunk 16]
```diff
@@ -0,0 +1,251 @@
+import { Texture, DynamicTexture, StandardMaterial, Color3, PBRMaterial, 
+  Scene as babylScene, Material as babylMaterial, GlowLayer } from '@babylonjs/core';
+
+import Material from '../../state/State/Scene/Material';
+import { Color } from '../../state/State/Scene/Color';
+import Patch from "../../util/redux/Patch";
+
+
+
+export const createMaterial = (id: string, material: Material, bScene_: babylScene) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/MaterialBinding.ts:10]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/MaterialBinding.ts [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+        }
+      }
+      bMaterial = pbr;
+      break;
+    }
+  }
+  return bMaterial;
+};
+
+export const updateMaterialBasic = (bMaterial: StandardMaterial, material: Patch.InnerPatch<Material.Basic>, bScene_: babylScene) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/MaterialBinding.ts:125]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/MaterialBinding.ts [hunk 18]
```diff
@@ -0,0 +1,251 @@
+import { Texture, DynamicTexture, StandardMaterial, Color3, PBRMaterial,
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/MaterialBinding.ts:1]: File name should be `camelCase`.

File: src/BabylonBindings/SceneObjects/MaterialBinding.ts [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+          bMaterial.diffuseTexture = new Texture(color.next.uri, bScene_);
+        }
+        break;
+      }
+    }
+  }
+  return bMaterial;
+};
+
+export const updateMaterialPbr = (bMaterial: PBRMaterial, material: Patch.InnerPatch<Material.Pbr>, bScene_: babylScene) => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/MaterialBinding.ts:154]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/ObjectBinding.ts [hunk 20]
```diff
@@ -0,0 +1,165 @@
+import { TransformNode, AbstractMesh, CreateBox, CreateSphere, CreateCylinder,
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/ObjectBinding.ts:1]: `camelCase`.

File: src/BabylonBindings/SceneObjects/ObjectBinding.ts [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+    return undefined;
+  }
+  const ret: Vector4[] = [];
+  for (let i = 0; i + 1 < faceUvs.length; i += 2) {
+    ret.push(new Vector4(faceUvs[i].x, faceUvs[i].y, faceUvs[i + 1].x, faceUvs[i + 1].y));
+  }
+  return ret;
+};
+
+export const buildGeometry = async (name: string, geometry: Geometry, bScene_: babylScene, faceUvs?: RawVector2[]): Promise<FrameLike> => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/ObjectBinding.ts:36]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/ObjectBinding.ts [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+  } else {
+    const children = ret.getChildren(c => c instanceof AbstractMesh) as Mesh[];
+    const mesh = Mesh.MergeMeshes(children, true, true, undefined, false, true);
+    mesh.visibility = 1;
+    ret = mesh;
+  }
+  return ret;
+};
+
+export const createObject = async (node: Node.Obj, nextScene: Scene, parent: babylNode, bScene_: babylScene): Promise<babylNode> => {
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/ObjectBinding.ts:141]: The `_` suffix is only for private class members.

File: src/BabylonBindings/SceneObjects/ObjectBinding.ts [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+    return null;
+  }
+  const ret = await buildGeometry(node.name[LocalizedString.EN_US], geometry, bScene_, node.faceUvs);
+  if (!node.visible) {
+    apply(ret, m => m.isVisible = false);
+  }
+  if (node.material) {
+    const material = createMaterial(node.name[LocalizedString.EN_US], node.material, bScene_);
+    apply(ret, m => m.material = material);
+  }
```
Comment (bmcdorman) [inline src/BabylonBindings/SceneObjects/ObjectBinding.ts:155]: The `_` suffix is only for private class members.

File: src/util/math/UnitMath.ts [hunk 24]
```diff
@@ -1,33 +1,33 @@
 // Contains math types that are tagged with units
```
Comment (bmcdorman) [inline src/util/math/UnitMath.ts:1]: make this file `camelCase` or retain current name.

File: src/util/math/UnitMath.ts [hunk 25]
```diff
@@ -1,33 +1,33 @@
...
+  RawAxisAngle,
+  RawReferenceFrame,
+  RawQuaternion,
 } from './math';
 
 import { TransformNode as BabylonTransformNode } from '@babylonjs/core/Meshes/transformNode';
 import { AbstractMesh as BabylonAbstractMesh } from '@babylonjs/core/Meshes/abstractMesh';
 
-export interface Vector2 {
+export interface Vector2wUnits {
```
Comment (bmcdorman) [inline src/util/math/UnitMath.ts:17]: What is a `2w`?

File: src/util/math/UnitMath.ts [hunk 26]
```diff
@@ -149,51 +149,51 @@ export namespace Vector3 {
...
 }
 
 
 
-export namespace Rotation {
+export namespace RotationwUnits {
   export type Type = 'euler' | 'axis-angle';
 
-  export interface Euler {
+  export interface EulerwUnits {
```
Comment (bmcdorman) [inline src/util/math/UnitMath.ts:172]: What is a `w`? Why doesn't it obey camel case?

File: src/util/math/UnitMath.ts [hunk 27]
```diff
@@ -263,65 +263,65 @@ export namespace Rotation {
...
+  export const slerp = (lhs: RotationwUnits, rhs: RotationwUnits, t: number, newType: Type = 'euler'): RotationwUnits => fromRawQuaternion(RawQuaternion.slerp(toRawQuaternion(lhs), toRawQuaternion(rhs), t), newType);
 }
 
-export type Rotation = Rotation.Euler | Rotation.AxisAngle;
+export type RotationwUnits = RotationwUnits.EulerwUnits | RotationwUnits.AxisAngle;
 
-export interface ReferenceFrame {
-  position?: Vector3;
-  orientation?: Rotation;
+export interface ReferenceFramewUnits {
```
Comment (bmcdorman) [inline src/util/math/UnitMath.ts:294]: Again, this `w` makes no sense.

File: src/util/redux/deepNeq.ts [hunk 28]
```diff
@@ -0,0 +1,115 @@
+// Disable eslint rule - this file deals heavily with "object" type
+/* eslint-disable @typescript-eslint/ban-types */
+
+/** A utility for deeply comparing two values (be they primitives, arrays, 
+ * or objects) to check if they are not equal, handling nested structures.
+ * Array Handling: If value0 is an array, it checks if value1 is also an array. 
+ *  If value1 is not an array, or if they are arrays but not deeply 
+ *  equal (checked by deepArrayNeq), it returns true (meaning they are not equal). 
+ *  If they are both arrays and deeply equal, it returns false.
+Type Comparison: Compares the types of value0 and value1. If they are of different
```
Comment (bmcdorman) [inline src/util/redux/deepNeq.ts:10]: This comment is malformed. Every line should start with `*`

File: src/util/redux/deepNeq.ts [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+  
+  if (value0 !== value1) return true;
+
+  return false;
+};
+
+
+/**
+ * Purpose: Compares two arrays to determine if they are not deeply equal.
+    Length Check: First, it checks if the arrays have different lengths. If so, it returns true.
```
Comment (bmcdorman) [inline src/util/redux/deepNeq.ts:58]: Malformed

File: src/util/redux/deepNeq.ts [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+  for (let i = 0; i < length; ++i) {
+    if (deepNeq(_0[i], _1[i])) return true;
+  }
+
+  return false;
+};
+
+/**
+*     Purpose: Compares two objects to determine if they are not deeply equal.
+  Key Length Check: Compares the number of keys in each object. If they have
```
Comment (bmcdorman) [inline src/util/redux/deepNeq.ts:78]: Malformed

File: src/state/reducer/robots.ts [hunk 31]
```diff
@@ -1,6 +1,6 @@
 import Robot from "../State/Robot";
 import Async from "../State/Async";
-import * as ROBOTS from '../../robots';
+import * as ROBOTS from '../../SimulatorDefinitions/robots';
```
Comment (bmcdorman) [inline src/state/reducer/robots.ts:3]: why is the `PascalCase`? This pattern is not used anywhere in the code base.

File: src/state/reducer/documentation.ts [hunk 32]
```diff
@@ -1,7 +1,7 @@
 import { DocumentationState } from '../State';
 import DocumentationLocation from '../State/Documentation/DocumentationLocation';
-import construct from '../../util/construct';
-import { Size } from '../../components/Widget';
+import construct from '../../util/redux/construct';
+import { Size } from '../../components/Interface/Widget';
```
Comment (bmcdorman) [inline src/state/reducer/documentation.ts:4]: This shouldn't be `Interface`, it should be `interface`.

File: src/state/reducer/challengeCompletions.ts [hunk 33]
```diff
@@ -1,17 +1,17 @@
...
 import ChallengeCompletion, { AsyncChallengeCompletion, ChallengeCompletionBrief } from '../State/ChallengeCompletion';
 import PredicateCompletion from '../State/ChallengeCompletion/PredicateCompletion';
 import Scene from '../State/Scene';
-import Dict from '../../Dict';
+import Dict from '../../util/objectOps/Dict';
 import { OuterObjectPatch } from 'symmetry/dist';
-import ProgrammingLanguage from '../../ProgrammingLanguage';
-import { ReferenceFrame } from '../../unit-math';
+import ProgrammingLanguage from '../../programming/ProgrammingLanguage';
+import { ReferenceFramewUnits } from '../../util/math/UnitMath';
```
Comment (bmcdorman) [inline src/state/reducer/challengeCompletions.ts:14]: What is this `w`?

File: src/state/State/index.ts [hunk 34]
```diff
@@ -1,6 +1,6 @@
-import { Size } from '../../components/Widget';
+import { Size } from '../../components/Interface/Widget';
```
Comment (bmcdorman) [inline src/state/State/index.ts:1]: Folder name shouldn't be `Interface`, it should be `interface`.

File: src/util/ParseMessages.ts [hunk 35]
```diff
@@ -1,4 +1,6 @@
-// Parses clang error messages
+/**
```
Comment (bmcdorman) [inline src/util/ParseMessages.ts:1]: Do not use `PascalCase`.

**Review by bmcdorman — CHANGES_REQUESTED** [review]
There are a lot of style changes here that do not align with the rest of the codebase. I'd love to see these fixed.

File: src/AbstractRobot/Motor.ts [hunk 36]
```diff
@@ -1,3 +1,31 @@
...
+  * position: This is a numeric property that represents the current position of the motor.
+  * pwm: This is another numeric property, which represents the pulse-width modulation value for the motor, 
+      a common method for controlling the power delivered to an electronic device like a motor.
+  * done: This boolean property indicates whether the motor has reached its goal position or speed.
+  * positionGoal and speedGoal: These numeric properties represent the target position and speed for the motor.
+  * kP, kI, and kD: These properties are the coefficients for a PID controller, a control loop mechanism commonly used in systems like this to control the motor's position and speed.
+  
+  The Motor namespace contains two enumerations, Direction and Mode, and a constant object NIL.
+
+  The Direction enumeration defines four possible directions for the motor: Idle, Forward, Backward, and Brake.
```
Comment (bmcdorman) [inline src/AbstractRobot/Motor.ts:19]: Comment needs `*`

File: src/App.tsx [hunk 37]
```diff
@@ -8,10 +8,10 @@ import Dashboard from './pages/Dashboard';
 import Tutorials from './pages/Tutorials';
 import Root from './components/Root';
 import { connect } from 'react-redux';
-import ChallengeRoot from './components/ChallengeRoot';
+import ChallengeRoot from './components/Challenge/ChallengeRoot';
 import DocumentationWindow from './components/documentation/DocumentationWindow';
 import { State as ReduxState } from './state';
-import { DARK } from './components/theme';
+import { DARK } from './components/Constants/theme';
```
Comment (bmcdorman) [inline src/App.tsx:14]: Do not use `PascalCase` for things that aren't classes.

**Review by tcorbly — APPROVED** [review]

**Review by tcorbly — APPROVED** [review]

**Pull request merged**